src/Nethermind/Nethermind.TxPool/Collections/SortedPool.cs

```diff
@@ -18,101 +18,104 @@
 using System.Collections.Generic;
 using System.Linq;
 using System.Runtime.CompilerServices;
+using Nethermind.Core.Collections;
 
 namespace Nethermind.TxPool.Collections
 {
-    public class SortedPool<TKey, TValue>
+    public class SortedPool<TKey, TValue, TGroup>
     {
         private readonly int _capacity;
-        protected readonly Comparison<TValue> Comparison;
-        protected readonly Dictionary<TKey, LinkedListNode<KeyValuePair<TKey, TValue>>> CacheMap;
-        protected readonly LinkedList<KeyValuePair<TKey, TValue>> LruList;
+        private readonly Func<TValue, TGroup> _groupMapping;
+        private readonly IDictionary<TGroup, ICollection<TValue>> _buckets;
+        private readonly DictionarySortedSet<TValue, TKey> _sortedValues;
+        
+        protected readonly IDictionary<TKey, TValue> CacheMap;
+        protected readonly IComparer<TValue> Comparer;
 
-        public SortedPool(int capacity, Comparison<TValue> comparison)
+        public SortedPool(int capacity, IComparer<TValue> comparer, Func<TValue, TGroup> groupMapping)
         {
             _capacity = capacity;
-            Comparison = comparison;
-            CacheMap = new Dictionary<TKey, LinkedListNode<KeyValuePair<TKey, TValue>>>(); // do not initialize it at the full capacity
-            LruList = new LinkedList<KeyValuePair<TKey, TValue>>();
+            Comparer = comparer;
+            _groupMapping = groupMapping;
+            CacheMap = new Dictionary<TKey, TValue>(); // do not initialize it at the full capacity
+            _buckets = new Dictionary<TGroup, ICollection<TValue>>();
+            _sortedValues = new DictionarySortedSet<TValue, TKey>(comparer);
         }
 
         public int Count => CacheMap.Count;
 
         [MethodImpl(MethodImplOptions.Synchronized)]
         public TValue[] GetSnapshot()
         {
-            return LruList.Select(i => i.Value).ToArray();
+            lock (CacheMap)
+            {
+                return _buckets.SelectMany(b => b.Value).ToArray();
+            }
         }
-
+        
         [MethodImpl(MethodImplOptions.Synchronized)]
-        public TValue TakeFirst()
+        public IDictionary<TGroup, TValue[]> GetBucketSnapshot()
         {
```

Comment:
why locking inside a synchronized method?

Reply:
ooooh, you did not notice it is all synchronized?

Reply:
easy to miss

Reply:
error, didn't think of that